File: common/src/main/java/net/shiroha233/roadweaver/features/bridge/BuoyMarkerPlanner.java
package net.shiroha233.roadweaver.features.bridge;

import net.minecraft.core.BlockPos;

import java.util.List;

public final class BuoyMarkerPlanner {
    private BuoyMarkerPlanner() {}

    private static double dist2d(BlockPos a, BlockPos b) {
        long dx = (long) b.getX() - a.getX();
        long dz = (long) b.getZ() - a.getZ();
        return Math.sqrt((double) dx * dx + (double) dz * dz);
    }

    public static boolean[] markersForBridgeRanges(List<BlockPos> middlePositions, List<int[]> bridgeRanges, int intervalBlocks) {
        if (middlePositions == null || middlePositions.isEmpty()) return new boolean[0];
        int n = middlePositions.size();
        boolean[] res = new boolean[n];
        if (bridgeRanges == null || bridgeRanges.isEmpty()) return res;

        int interval = Math.max(1, intervalBlocks);
        for (int[] r : bridgeRanges) {
            if (r == null || r.length < 2) continue;
            int s = Math.max(0, r[0] + 1);
            int e = Math.min(n - 1, r[1] - 1);
            if (s > e) continue;

            double acc = 0.0;
            double next = 0.0;
            for (int idx = s; idx <= e; idx++) {
                if (idx > s) {
                    acc += dist2d(middlePositions.get(idx - 1), middlePositions.get(idx));
                }
                if (acc + 1e-6 >= next) {
                    res[idx] = true;
                    next += interval;
                }
            }
        }
        return res;
    }

    public static boolean[] markersForMask(List<BlockPos> middlePositions, boolean[] mask, int intervalBlocks) {
        if (middlePositions == null || middlePositions.isEmpty()) return new boolean[0];
        int n = middlePositions.size();
        boolean[] res = new boolean[n];
        if (mask == null || mask.length != n) return res;

        int interval = Math.max(1, intervalBlocks);
        int idx = 0;
        while (idx < n) {
            while (idx < n && !mask[idx]) idx++;
            if (idx >= n) break;
            int start = idx;
            while (idx < n && mask[idx]) idx++;
            int end = idx - 1;

            double acc = 0.0;
            double next = 0.0;
            for (int i = start; i <= end; i++) {
                if (i > start) {
                    acc += dist2d(middlePositions.get(i - 1), middlePositions.get(i));
                }
                if (acc + 1e-6 >= next) {
                    res[i] = true;
                    next += interval;
                }
            }
        }
        return res;
    }
}

File: common/src/main/java/net/shiroha233/roadweaver/features/path/bridge/BuoyMarkerPlanner.kt
package net.shiroha233.roadweaver.features.path.bridge

import net.minecraft.core.BlockPos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

object BuoyMarkerPlanner {
    private fun dist2d(a: BlockPos, b: BlockPos): Double {
        val dx = (b.x.toLong() - a.x.toLong())
        val dz = (b.z.toLong() - a.z.toLong())
        return sqrt(dx.toDouble() * dx.toDouble() + dz.toDouble() * dz.toDouble())
    }

    @JvmStatic
    fun markersForBridgeRanges(
        middlePositions: List<BlockPos>?,
        bridgeRanges: List<IntArray>?,
        intervalBlocks: Int
    ): BooleanArray {
        if (middlePositions.isNullOrEmpty()) return BooleanArray(0)
        val n = middlePositions.size
        val res = BooleanArray(n)
        if (bridgeRanges.isNullOrEmpty()) return res

        val interval = max(1, intervalBlocks)
        for (r in bridgeRanges) {
            if (r.size < 2) continue
            val s = max(0, r[0] + 1)
            val e = min(n - 1, r[1] - 1)
            if (s > e) continue

            var acc = 0.0
            var next = 0.0
            for (idx in s..e) {
                if (idx > s) {
                    acc += dist2d(middlePositions[idx - 1], middlePositions[idx])
                }
                if (acc + 1e-6 >= next) {
                    res[idx] = true
                    next += interval
                }
            }
        }
        return res
    }

    @JvmStatic
    fun markersForMask(
        middlePositions: List<BlockPos>?,
        mask: BooleanArray?,
        intervalBlocks: Int
    ): BooleanArray {
        if (middlePositions.isNullOrEmpty()) return BooleanArray(0)
        val n = middlePositions.size
        val res = BooleanArray(n)
        if (mask == null || mask.size != n) return res

        val interval = max(1, intervalBlocks)
        var idx = 0
        while (idx < n) {
            while (idx < n && !mask[idx]) idx++
            if (idx >= n) break
            val start = idx
            while (idx < n && mask[idx]) idx++
            val end = idx - 1

            var acc = 0.0
            var next = 0.0
            for (i in start..end) {
                if (i > start) {
                    acc += dist2d(middlePositions[i - 1], middlePositions[i])
                }
                if (acc + 1e-6 >= next) {
                    res[i] = true
                    next += interval
                }
            }
        }

        return res
    }
}
